spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -207,68 +208,87 @@ private StringBuilder wrapAsSubSelect(String sql) {
 
 	@Override
 	public void insert(Object object) {
-		applyMutationsUsingEntity(this.mutationFactory::insert, object);
+		applyMutations(this.mutationFactory.insert(object));
+	}
+
+	@Override
+	public void insertAll(Iterable objects) {
+		applyMutations(
+				getMutationsForMultipleObjects(objects, this.mutationFactory::insert));
 	}
 
 	@Override
 	public void update(Object object) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object, null);
+		applyMutations(this.mutationFactory.update(object, null));
+	}
+
+	@Override
+	public void updateAll(Iterable objects) {
+		applyMutations(
+				getMutationsForMultipleObjects(objects,
+						x -> this.mutationFactory.update(x, null)));
 	}
 
 	@Override
 	public void update(Object object, String... includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object,
+		applyMutations(
+				this.mutationFactory.update(object,
 				includeColumns.length == 0 ? null
-						: Optional.of(new HashSet<>(Arrays.asList(includeColumns))));
+						: Optional.of(new HashSet<>(Arrays.asList(includeColumns)))));
 	}
 
 	@Override
 	public void update(Object object, Optional<Set<String>> includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::update, object, includeColumns);
+		applyMutations(this.mutationFactory.update(object, includeColumns));
 	}
 
 	@Override
 	public void upsert(Object object) {
-		applyMutationsTwoArgs(this.mutationFactory::upsert, object, null);
+		applyMutations(this.mutationFactory.upsert(object, null));
+	}
+
+	@Override
+	public void upsertAll(Iterable objects) {
+		applyMutations(
+				getMutationsForMultipleObjects(objects,
+						x -> this.mutationFactory.upsert(x, null)));
 	}
 
 	@Override
 	public void upsert(Object object, String... includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::upsert, object,
+		applyMutations(
+				this.mutationFactory.upsert(object,
 				includeColumns.length == 0 ? null
-						: Optional.of(new HashSet<>(Arrays.asList(includeColumns))));
+						: Optional.of(new HashSet<>(Arrays.asList(includeColumns)))));
 	}
 
 	@Override
 	public void upsert(Object object, Optional<Set<String>> includeColumns) {
-		applyMutationsTwoArgs(this.mutationFactory::upsert, object, includeColumns);
+		applyMutations(this.mutationFactory.upsert(object, includeColumns));
 	}
 
 	@Override
 	public void delete(Object entity) {
-		applyMutationUsingEntity(
-				this.mutationFactory::delete, entity);
+		applyMutations(Collections.singletonList(this.mutationFactory.delete(entity)));
 	}
 
 	@Override
-	public void delete(Class entityClass, Key key) {
-		applyMutationTwoArgs(
-				this.mutationFactory::delete,
-				entityClass, key);
+	public void deleteAll(Iterable objects) {
+		applyMutations(
+				(Collection<Mutation>) StreamSupport.stream(objects.spliterator(), false)
```

Comment:
Can't use `getMutationsForMultipleObjects` here?

Reply:
That method requires each element produces a collection of mutations, which isn't the case for `delete` because of relying on the cascade for delete. as a result, we would need to wrap each delete mutation into a singleton list, and the end result doesn't look better than directly making the list of mutations as shown here.